tests/EStimLibrary.UnitTests/Core/HardwareInterfaces/LeadTests.cs

```diff
@@ -0,0 +1,618 @@
+                new object[]
+                {
+                    1, false, new Lead(new SortedSet<int> { 1, 2 },
+                        new SortedSet<int> { 3, 4, 5 }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { 3, 4, 5 }, true
+                },
+                new object[]
+                {
+                    2, false, new Lead(new SortedSet<int> { 1, 2 },
+                        new SortedSet<int> { 3, 4, 5, 6, 7 }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { 3, 4, 5, 6, 7 }, true
+                },
+                new object[]
+                {
+                    2, false, new Lead(new SortedSet<int> { 2 },
+                        new SortedSet<int> { }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { }, true
+                },
+                new object[]
+                {
+                    2, false, new Lead(new SortedSet<int> { 1, 2 },
+                        new SortedSet<int> { 1 }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { 1 }, true
+                },
+                // Testing with a contact as the input which is not connected to the lead
+                // We expect to get false and an unmodified output set
+                new object[]
+                {
+                    1, false, new Lead(new SortedSet<int> { 2 },
+                        new SortedSet<int> { 3, 4, 5 }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { 3, 4, 5 }, false
+                },
+                new object[]
+                {
+                    2, false, new Lead(new SortedSet<int> { },
+                        new SortedSet<int> { 3, 4, 5, 6, 7 }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { 3, 4, 5, 6, 7 }, false
+                },
+                new object[]
+                {
+                    2, false, new Lead(new SortedSet<int> { 9 },
+                        new SortedSet<int> { }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { }, false
+                },
+                new object[]
+                {
+                    2, false, new Lead(new SortedSet<int> { 5, 6, 7, 8 },
+                        new SortedSet<int> { 1 }, Constants.CurrentDirection.SINK),
+                    new SortedSet<int> { 1 }, false
+                }
+            };
+        }
+
+        // Test for GetConnectedContacts
+        [Theory]
+        [MemberData(nameof(GetConnectedContactsTestData))]
+        public void GetConnectedContacts_ShouldReturnExpectedResults(int id, bool searchIsAContact,
+            Lead lead, SortedSet<int> expectedContacts, bool expectedResult)
+        {
+            // Act
+            var result = lead.GetConnectedContacts(id, searchIsAContact, out var connectedContacts);
+
+            // Assert
+            Assert.Equal(expectedResult, result);
```

Comment:
like GetConnectedOutputs: Add an assertion that the lead's actual output and contact sets are not modified from before to after the GetConnectedContacts call (check address and data values). May need to store shallow (for address checking) and deep (for value checking) copies of the sets prior to the method call, then use those as the expected values in the asserts, comparing against the lead's properties at that time